### PR TITLE
feat: remove icu_datagen

### LINKED ISSUE #99: [REQ] Automatically download missing locale tags
At the moment, an error is returned if the locale-tag set is not supported by the default locales, and no locale-file is given with download option, or that already supports it.

It would be better user experience if we automatically try to download the required locale-tags.
If a locale-file is given, but does not yet exist, it will be used as download destination.
Otherwise, a locale-file, with all required locale-tags set as filename, should be created inside a `.unimarkup` folder in the home directory of the user.

The `download` flag would become unnecessary, and must be removed.

### LINKED ISSUE #102: [REQ] Replace icu_datagen dynamic local download with icu rust module
For grapheme segmentation, we use the icu crate.
This crate needs locale data, and at the moment we have a `.postcard` file for a few default locales, but for other locales,
we used icu_datagen to download this data if needed.
While this seemed like a good solution, it turns out that the required data for all locales is pretty much the same size compared to the one supporting only the default ones.

To verify this, I created a `.postcard` file with `icu4x-datagen --keys-for-bin .\target\debug\unimarkup.exe --locales full --format blob --out full_icu_blob.postcard`.
This file is about 14KB.

To get this small file, I had to remove all code using `icu_datagen` , because with the code as it is right now the file was ~22MB.
I assume it is because `--keys-for-bin` found `icu_datagen` in the binary, and this basically requires all data :D

Removing `icu_datagen` as dependency significantly reduces our dependencies.
We can also use the `mod` format of icu to get the data directly instead of deserializing it, which means we do not need the serde feature.

In addition, we should wrap the initial provider in `LocaleFallbackprovider` to allow locale fallback handling.
e.g. `en-US` to `en` (see: https://github.com/unicode-org/icu4x/blob/main/docs/tutorials/data_management.md)

### PR DESCRIPTION
**Pull request by mhatzl**
closes #102 closes #99 (makes #99 obsolet)

This PR removes dependency for the icu_datagen crate as described in issue #102 

Instead of using blob data, I decided to use the `mod` option offered by icu.
This embeds the needed data as rust module, which I generated using `icu4x-datagen --keys-for-bin .\target\debug\unimarkup.exe --locales full --format mod --out ./commons/src/scanner/icu_data`

The module option allows to use the `UnstableProvider` of icu, which offers better performance.
`Unstable`, because the icu crate versions, and the generated module must match.
In [this](https://github.com/unicode-org/icu4x/blob/main/docs/tutorials/data_management.md) tutorial, they mention that the module should be generated using a build pipeline to ensure a correct version.
For now, I set all icu crates to a fixed version, which should also be fine, but we might want to create a build script to handle this at some point.

With a concrete provider, the `Scanner` struct has no need for a generic anymore.
It is also not needed anymore to pass a provider to create a new scanner.

### CODE REVIEW

File: commons/src/config/preamble.rs [hunk 1]
```diff
@@ -54,208 +50,21 @@ pub struct I18n {
...
     #[arg(long, value_parser = parse_to_hashset::<Locale>, required = false, default_value = "")]
     #[serde(with = "locale::serde::multiple")]
-    pub langs: HashSet<Locale>,
-
-    #[arg(long = "locales-file")]
-    pub locales_file: Option<PathBuf>,
-
-    #[arg(long = "download-locales")]
-    pub download: bool,
+    pub output_langs: HashSet<Locale>,
```
Comment (nfejzic) [inline commons/src/config/preamble.rs:53]: Why the change to `output_langs`? Does the flag used also change?

File: commons/src/scanner/mod.rs [hunk 2]
```diff
@@ -56,7 +34,18 @@ where
     }
 }
 
-impl<P> Scanner<P> {
+impl Scanner {
+    pub fn try_new() -> Result<Self, SegmenterError> {
+        let icu_data_provider = IcuDataProvider;
+        let fallback_provider = LocaleFallbackProvider::try_new_unstable(icu_data_provider)?;
+        let segmenter = GraphemeClusterSegmenter::try_new_unstable(&fallback_provider)?;
```
Comment (nfejzic) [inline commons/src/scanner/mod.rs:41]: Is it known under which circumstances does this fail? Maybe we can change this function to be infallible (return `Scanner` directly without wrapping it into `Result`) now that we use module as data and can ensure compatibility.

File: commons/Cargo.toml [hunk 3]
```diff
@@ -19,11 +19,10 @@ serde.workspace = true
 serde_json.workspace = true
 serde_yaml.workspace = true
 once_cell = { workspace = true, optional = true }
-icu_locid = "=1.2.0"
-icu_segmenter = "=1.2.0"
-icu_provider = { version = "=1.2.0" }
-icu_provider_blob = "=1.2.0"
-icu_datagen = { version = "=1.2.5", default-features=false, features = [ "networking" ] }
+icu = "=1.2.0"
```
Comment (nfejzic) [inline commons/Cargo.toml:22]: I assume this is required by the module based icu data? It would be nice if we could only add dependencies that we use.

**Review by nfejzic — APPROVED** [review]
There are some nitpicks that might be ever so slightly more ergonomic IF we find more ergonomic solutions. In general, I'm very happy with this!

File: commons/src/config/preamble.rs (hunk 1 above)
Comment (mhatzl) [reply commons/src/config/preamble.rs:53]: this flag is now for output only. before, this flag had a kind of double use, because we also derived the locales that might appear in the document. now, it is only to indicate the languages that a document should be translated to. I will also update this in the spec accordingly

File: commons/src/scanner/mod.rs (hunk 2 above)
Comment (mhatzl) [reply commons/src/scanner/mod.rs:41]: I am not sure. 
it definitely fails if the generated module and used crate versions are not in-sync, but there could also be other reasons. would there be much benefit to make scanner infallible?
I assume making it infallible means panicking if it fails, but in our case if creating the scammer fails, we kind of also stop everything 🤔 so there is not that much of a difference?

File: commons/Cargo.toml (hunk 3 above)
Comment (mhatzl) [reply commons/Cargo.toml:22]: yes. the module requires icu directly. i removed all dependencies that are included in the icu to keep it minimal

File: commons/src/scanner/mod.rs (hunk 2 above)
Comment (nfejzic) [reply commons/src/scanner/mod.rs:41]: If that's the only reason it could fail, then I would say it would almost be better to simply panic at this point. That is one of the cases where we can't recover from the error in any meaningful way, so crashing it is better option to catch the problem as soon as possible. 

It would be best if we can fail at compile time, so that failure is really ruled out if we ship a compiled binary. But I'm not sure if that's possible at all.

Comment (mhatzl) [reply commons/src/scanner/mod.rs:41]: maybe using once_cell to create a static scanner? 😅

Comment (nfejzic) [reply commons/src/scanner/mod.rs:41]: Alternative to panic would be to log errors inside of this function and then call [`std::process::exit(1)`](https://doc.rust-lang.org/std/process/fn.exit.html) or something like that.

Comment (nfejzic) [reply commons/src/scanner/mod.rs:41]: > maybe using once_cell to create a static scanner? 😅

I don't think that would help. We would need the actual provider and segmenter to be constructed at compile time, but their functions are not `const`, so that wont work.

Comment (nfejzic) [reply commons/src/scanner/mod.rs:41]: I think that would be a nice solution, with error logging. We could also include some debug logs to provide more information on what's gone wrong. 

However, I suggest we make an issue for that one and investigate a little more. It might be the case that these functions return error in more circumstances than what we assume right now. This is fine for now 👍🏻
